### PR TITLE
OGC API Records / API improvements

### PR DESCRIPTION
**Pull request by fxprunayre**
## /items support DCAT output.

Before that change, DCAT output was only provided per records and not for search results.

Downloading a large catalog in DCAT format

```shell
curl -v "http://localhost:9901/collections/main/items?f=dcat&limit=12000" > rdf.xml
```

Made some changes to DCAT mapper to not fail on invalid dates.

Add DataService mapping for services.

Check DCAT-AP validation using https://www.itb.ec.europa.eu/shacl/dcat-ap/upload

image

### CODE REVIEW

**Comment by fgravin** [comment]
Thanks @fxprunayre 
Does it replace the old https://www.datagrandest.fr/geonetwork/srv/fre/rdf.search service ? Which means GN could be harvested by data.gouv.fr ?

**Comment by fxprunayre** [comment]
> Does it replace the old

It could probably be considered as an alternative.

BTW data.gouv.fr should probably rather support harvesting GeoNetwork 3 or 4 using CSW and its DCAT output (see https://github.com/etalab/doc.data.gouv.fr/commit/f8a0b32f19f46c22c30add1bfa5cda248861531e#r75959289) which is probably the more standardized way to get the catalogue content as DCAT (that we are providing for years now - see https://github.com/geonetwork/core-geonetwork/pull/1407). `rdf.search` was more a quite old internal/experimental API (see https://trac.osgeo.org/geonetwork/wiki/proposals/DCATandRDFServices). OGC API Records is still a draft.

**Comment by hadfoo** [comment]
Hello,

I have cloned your branch. Thank you for your job.
I've noticed that the conversion to dcat is still not developed for large catalogs.
I've done this locally. Are you interested in my contributions ?

File: modules/library/common-search/src/main/java/org/fao/geonet/common/search/processor/impl/DcatCatalogResponseProcessorImpl.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    JsonParser parser = parserForStream(streamFromServer);
+    new ResponseParser().matchHits(parser, generator, doc -> {
+      writeItem(generator, streamToClient, doc);
+    }, false);
+
+    generator.writeEndElement();
+    generator.writeEndDocument();
+    generator.flush();
+    generator.close();
```
Comment (josegar74) [inline modules/library/common-search/src/main/java/org/fao/geonet/common/search/processor/impl/DcatCatalogResponseProcessorImpl.java:76]: Should be done in a `finally` block?

File: modules/library/common-search/src/main/java/org/fao/geonet/common/search/processor/impl/GeoJsonResponseProcessorImpl.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          }
+        }
+      }, false);
+    }
+    generator.writeEndArray();
+    //    generator.writeNumberField("took", 0);
+    generator.writeNumberField("size", responseParser.total);
+    generator.writeEndObject();
+    generator.flush();
+    generator.close();
```
Comment (josegar74) [inline modules/library/common-search/src/main/java/org/fao/geonet/common/search/processor/impl/GeoJsonResponseProcessorImpl.java:74]: Should be done in a `finally` block?

**Review by josegar74 — APPROVED** [review]

**Comment by fxprunayre** [comment]
> I've noticed that the conversion to dcat is still not developed for large catalogs.

Not sure what you mean by that. This PR provides option to dump a full catalog in DCAT format.

> I've done this locally. Are you interested in my contributions ?

Contributions are welcomed.

**Pull request merged**